tests/h2-tests/tests/push_promise.rs

```diff
@@ -189,9 +255,58 @@ fn recv_push_promise_over_max_header_list_size() {
 }
 
 #[test]
-#[ignore]
-fn recv_push_promise_with_unsafe_method_is_stream_error() {
-    // for instance, when :method = POST
+fn recv_invalid_push_promise_headers_is_stream_protocol_error() {
+    // Unsafe method or content length is stream protocol error
+    let _ = ::env_logger::try_init();
+
+    let (io, srv) = mock::new();
+    let mock = srv.assert_client_handshake()
+        .unwrap()
+        .recv_settings()
+        .recv_frame(
+            frames::headers(1)
+                .request("GET", "https://http2.akamai.com/")
+                .eos(),
+        )
+        .send_frame(frames::headers(1).response(404))
+        .send_frame(frames::push_promise(1, 2).request("POST", "https://http2.akamai.com/style.css"))
+        .send_frame(
+            frames::push_promise(1, 4)
+                .request("GET", "https://http2.akamai.com/style.css")
+                .field(http::header::CONTENT_LENGTH, 1)
+        )
+        .send_frame(
+            frames::push_promise(1, 6)
+                .request("GET", "https://http2.akamai.com/style.css")
+                .field(http::header::CONTENT_LENGTH, 0)
+        )
+        .send_frame(frames::headers(1).response(404).eos())
+        .recv_frame(frames::reset(2).protocol_error())
+        .recv_frame(frames::reset(4).protocol_error())
+        .send_frame(frames::headers(6).response(200).eos())
+        .close();
+
+    let h2 = client::handshake(io).unwrap().and_then(|(mut client, h2)| {
+        let request = Request::builder()
+            .method(Method::GET)
+            .uri("https://http2.akamai.com/")
+            .body(())
+            .unwrap();
+        let (mut resp, _) = client
+            .send_request(request, true)
+            .unwrap();
+        let check_pushed_request = resp.push_promises().and_then(|headers| {
+            headers.into_parts().1
+        });
+        let check_pushed_response = check_pushed_request
+            .collect().unwrap().map(|ps| {
+                // CONTENT_LENGTH = 0 is ok
+                assert_eq!(1, ps.len())
+            });
+        h2.drive(check_pushed_response)
+    });
+
+    h2.join(mock).wait().unwrap();
 }
 
 #[test]
```

Comment:
Oh I wanted to point out, this check isn't implemented yet. It wasn't immediately clear to me how.

> The server MUST include a value in the :authority pseudo-header field for which the server is authoritative (see Section 10.1). A client MUST treat a PUSH_PROMISE for which the server is not authoritative as a stream error (Section 5.4.2) of type PROTOCOL_ERROR.

Reply:
@carllerche

Reply:
I looked briefly into it, but it doesn't look possible to implement at the `h2` level because the library is decoupled from the real authority.

Thoughts 
@seanmonstar ?

Reply:
I would probably file a follow up issue. I believe requests and responses have the same issue right now.

Reply:
I filed: https://github.com/carllerche/h2/issues/323